### PR TITLE
#12133 Speed up twisted.web's HTTP/1.1 server by about 10%

### LINKED ISSUE #12133: Speed up twisted.web, part 1 of N
Going to try do the equivalent of what I did for twisted.web client: fix a bunch of small performance issues that add up. Initial investigation suggests there will be things to fix, e.g. various overhead from deprecated code paths.

Might be a series of issues, will see.

### PR DESCRIPTION
**Pull request by itamarst**
## Scope and purpose

Fixes #12133

Doing deprecations on module-level is expensive, turns out, because it adds lots of attribute-lookup overhead for _everything_, including non-deprecated code. And other speedups as well.

Before:

```
$ hyperfine --warmup 1 "python http_server_benchmark.py"
Benchmark 1: python http_server_benchmark.py
  Time (mean ± σ):      1.220 s ±  0.010 s    [User: 1.190 s, System: 0.030 s]
  Range (min … max):    1.207 s …  1.238 s    10 runs
```

After:

```
$ hyperfine --warmup 1 "python http_server_benchmark.py"
Benchmark 1: python http_server_benchmark.py
  Time (mean ± σ):      1.105 s ±  0.012 s    [User: 1.081 s, System: 0.024 s]
  Range (min … max):    1.096 s …  1.137 s    10 runs
```

## Contributor Checklist:

This process applies to *all* pull requests - no matter how small.
Have a look at [our developer documentation](https://docs.twisted.org/en/latest/core/development/dev-process.html) before submitting your Pull Request.

Below is a non-exhaustive list (as a reminder):

* The title of the PR should describe the changes and starts with the associated issue number, like “#1234 Brief description”.
* A release notes news fragment file was create in src/twisted/newsfragments/ (see: [Release notes fragments docs.](https://docs.twisted.org/en/latest/core/development/dev-process.html#release-notes-management))
* The automated tests were updated.
* Once all checks are green, request a review by leaving a comment that contains exactly the string `please review`.
  Our bot will trigger the review process, by applying the pending review label
  and requesting a review from the Twisted dev team.

### CODE REVIEW

File: src/twisted/web/http.py [hunk 1]
```diff
@@ -3145,11 +3136,9 @@ def dataReceived(self, data):
         using.
         """
         if self._negotiatedProtocol is None:
-            try:
-                negotiatedProtocol = self._channel.transport.negotiatedProtocol
-            except AttributeError:
-                # Plaintext HTTP, always HTTP/1.1
-                negotiatedProtocol = b"http/1.1"
+            negotiatedProtocol = getattr(
```
Comment (itamarst) [inline src/twisted/web/http.py:3139]: It's possible there's some edge case where this is an incompatible change?

File: src/twisted/web/resource.py [hunk 2]
```diff
@@ -355,7 +349,26 @@ def getChild(self, chnam, request):
         return self
 
 
-class _UnsafeNoResource(_UnsafeErrorPage):
+class _UnsafeErrorPage(_UnsafeErrorPageBase):
```
Comment (adiroiban) [inline src/twisted/web/resource.py:352]: I think that we can just remove this code.

We have this code available until 23.4

Based on [The First One’s Always Free](https://docs.twisted.org/en/stable/development/compatibility-policy.html#the-first-one-s-always-free), we already had than one release with this warning, so we can remove it.

File: src/twisted/python/deprecate.py [hunk 3]
```diff
@@ -47,7 +47,8 @@ def badProperty(self, value):
             '''
 
 
-To mark module-level attributes as being deprecated you can use::
+While it's best to avoid this as it adds performance overhead to *any* usage of
```
Comment (adiroiban) [inline src/twisted/python/deprecate.py:50]: I think that we can have this here.

Maybe also add a note that API deprecated in this way should be removed at the N+1 release cycle.

There is a separate PR to look at the overall update of the deprecation docs https://github.com/twisted/twisted/pull/11971

File: src/twisted/web/http.py [hunk 4]
```diff
@@ -385,7 +385,7 @@ def stringToDatetime(dateString):
 
     @type dateString: C{bytes}
     """
-    parts = nativeString(dateString).split()
+    parts = dateString.decode("ascii").split()
```
Comment (adiroiban) [inline src/twisted/web/http.py:388]: I think there are multiple places in which we use nativeString ... and other helpers for 2to3 migration.

I think that is ok to have this change here... but maybe have a separate task to completely remove the usage of nativeString and deprecate nativeString

File: src/twisted/web/http.py [hunk 5]
```diff
@@ -1363,19 +1363,15 @@ def _sanitize(val):
...
 
-    def setResponseCode(self, code, message=None):
+    def setResponseCode(self, code: int, message: Optional[bytes] = None) -> None:
         """
         Set the HTTP response code.
 
         @type code: L{int}
         @type message: L{bytes}
         """
-        if not isinstance(code, int):
```
Comment (adiroiban) [inline src/twisted/web/http.py:1373]: Is this a backward compatible change in terms of how this public API behaves and handles error cases ?

--------

I know that mypy will help with such errors and I agree that we should not have this error handling at runtime

------


I don't know we can raise a warning to inform on this incompatible change.


I think that what we can do is to remove this code and
add a newsfragment/12133.removal release note to let others know about this change

File: src/twisted/web/http.py [hunk 6]
```diff
@@ -2275,13 +2271,15 @@ def connectionMade(self):
         )
         self._networkProducer.registerProducer(self, True)
 
+    def dataReceived(self, data):
```
Comment (adiroiban) [inline src/twisted/web/http.py:2274]: thanks for the change.

I was expecting that this is already handled via the PolicyMixin ... byt it looks like it's not.

https://github.com/twisted/twisted/blob/10fc2f1cf0500f59497557d65b6751d473c40d85/src/twisted/protocols/policies.py#L620

File: src/twisted/web/http.py [hunk 7]
```diff
@@ -2427,12 +2425,7 @@ def headerReceived(self, line):
         if not self._maybeChooseTransferDecoder(header, data):
             return False
 
-        reqHeaders = self.requests[-1].requestHeaders
-        values = reqHeaders.getRawHeaders(header)
-        if values is not None:
-            values.append(data)
-        else:
-            reqHeaders.setRawHeaders(header, [data])
+        self.requests[-1].requestHeaders.addRawHeader(header, data)
```
Comment (adiroiban) [inline src/twisted/web/http.py:2428]: Thanks for the cleanup.

I guess that addRawHeader API was added later, and this is why we had the previous usage with get and set

File: src/twisted/web/http.py (hunk 1 above)
Comment (adiroiban) [reply src/twisted/web/http.py:3139]: The change looks ok to me.

Was this just a drive-by clenup, or this has a performance impact?

File: src/twisted/web/server.py [hunk 8]
```diff
@@ -53,20 +53,21 @@
 
 
 # backwards compatibility
-deprecatedModuleAttribute(
+@deprecated(
```
Comment (adiroiban) [inline src/twisted/web/server.py]: Can we remove this ?

**Review by adiroiban — APPROVED** [review]
Thanks for the changes.

They look good.

Things to consider before the merge:

* Add a release note that the TypeError is no longer raised by the public API

-------

We can have the deprecated code merged,
but I would say that we should just remove it.
I can do that in a separate PR.

File: src/twisted/web/http.py (hunk 1 above)
Comment (itamarst) [reply src/twisted/web/http.py:3139]: About a 2% performance improvement.

File: src/twisted/web/resource.py (hunk 2 above)
Comment (itamarst) [reply src/twisted/web/resource.py:352]: This is... pretty long-lived API I think? So I am reluctant to break existing applications that might rely on it.

Comment (itamarst) [reply src/twisted/web/resource.py:352]: Specifically the bit in the docs about "Removal should not be undertaken just to follow a timeline. Twisted should strive, as much as practical, not to break applications relying on it."

File: src/twisted/web/http.py (hunk 4 above)
Comment (itamarst) [reply src/twisted/web/http.py:388]: Yeah I did it here because it showed up in the profiling.

File: src/twisted/web/server.py (hunk 8 above)
Comment (itamarst) [reply src/twisted/web/server.py]: I guess 12 years is a generous amount of time. I'll do that.

**Comment by adiroiban** [comment]
@glyph @itamarst about performance testing.

I have approved the request for codspeed ... 

--------


I think that one easy way is to write performance tests, is to write them as trial test.

We can have a pluging to trial that will write a report about the test run.
The report can include the time it takes for the test to run with setup + test + teardown, and just the "inner" test.. all in xunit format.

This data can then be sent to Datadog and we can generate graphs and dashboards.
With datadog import tool, we can add extra attributes for each test.
Datadog offered a free account...as long as we put a banner on our website.

I have implemented this for my private project... but stopped using Datadog as it was too expensive.
I am now working to move the database handling to prometheus or influxdb .

--------

The problem that I can see, is having a reliable / reference run environemnt.
From my past experience, the GitHub Action VMs were not alwasy performing the same.

I can volunteer to host a bare-metal server, dedicated to performance testing, and we can guarantee to have the same results.

File: src/twisted/web/http.py (hunk 1 above)
Comment (adiroiban) [reply src/twisted/web/http.py:3139]: wow! I was not aware of this. Thanks again for the update

**Comment by itamarst** [comment]
I think the point of codspeed.io is that they provide a reliable runtime environment. We'll see how accurate that claim is.

**Pull request merged**